pkg/driver/node.go

```diff
@@ -0,0 +1,566 @@
+	notMnt, err := ns.mounter.IsLikelyNotMountPoint(mntPath)
+	if err != nil && !os.IsNotExist(err) {
+		if err := os.MkdirAll(mntPath, 0750); err != nil {
+			logrus.Errorf("failed to mkdir %s, error", mntPath)
+			return err
+		}
+	}
+
+	if !notMnt {
+		logrus.Infof("Volume %s has been mounted already at %v", req.GetVolumeId(), mntPath)
+		return nil
+	}
+
+	fsType := req.GetVolumeCapability().GetMount().GetFsType()
+	options := []string{}
+	mountFlags := req.GetVolumeCapability().GetMount().GetMountFlags()
+	options = append(options, mountFlags...)
+
+	err = ns.mounter.FormatAndMount(devicePath, mntPath, fsType, options)
+	if err != nil {
+		logrus.Errorf(
+			"Failed to mount iscsi volume %s [%s, %s] to %s, error %v",
+			req.GetVolumeId(), devicePath, fsType, mntPath, err,
+		)
+		return err
+	}
+	return nil
+}
+
+// NodePublishVolume publishes (mounts) the volume
+// at the corresponding node at a given path
+//
+// This implements csi.NodeServer
+func (ns *node) NodePublishVolume(
+	ctx context.Context,
+	req *csi.NodePublishVolumeRequest,
+) (*csi.NodePublishVolumeResponse, error) {
+
+	volumeID := req.GetVolumeId()
+	if len(volumeID) == 0 {
+		return nil, status.Error(codes.FailedPrecondition, "Volume ID not provided")
+	}
+
+	target := req.GetTargetPath()
+	if len(target) == 0 {
+		return nil, status.Error(codes.FailedPrecondition, "Target path not provided")
+	}
+
+	volCap := req.GetVolumeCapability()
+	if volCap == nil {
+		return nil, status.Error(codes.FailedPrecondition, "Volume capability not provided")
+	}
+
+	if !isValidVolumeCapabilities([]*csi.VolumeCapability{volCap}) {
+		return nil, status.Error(codes.InvalidArgument, "Volume capability not supported")
+	}
+
+	if err := ns.client.Set(); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	mountOptions := []string{"bind"}
+	if req.GetReadonly() {
+		mountOptions = append(mountOptions, "ro")
+	}
+	switch mode := volCap.GetAccessType().(type) {
+	case *csi.VolumeCapability_Block:
+		return &csi.NodePublishVolumeResponse{}, status.Error(codes.Unimplemented, "doesn't support block device provisioning")
+	case *csi.VolumeCapability_Mount:
+		if err := ns.nodePublishVolumeForFileSystem(req, mountOptions, mode); err != nil {
+			return nil, err
+		}
+	}
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func (ns *node) nodePublishVolumeForFileSystem(req *csi.NodePublishVolumeRequest, mountOptions []string, mode *csi.VolumeCapability_Mount) error {
+	target := req.GetTargetPath()
+	source := req.GetStagingTargetPath()
+	if m := mode.Mount; m != nil {
+		hasOption := func(options []string, opt string) bool {
+			for _, o := range options {
+				if o == opt {
+					return true
+				}
+			}
+			return false
+		}
+		for _, f := range m.MountFlags {
+			if !hasOption(mountOptions, f) {
+				mountOptions = append(mountOptions, f)
+			}
+		}
+	}
+
+	logrus.Infof("NodePublishVolume: creating dir %s", target)
+	if err := ns.mounter.MakeDir(target); err != nil {
```

Comment:
Set file permissions to be 000, so that when the volume goes into read-only and it is unmounted from the path to mount again, the app should not start writing/reading at the path without the mount point.

Reply:
Lets do a separate call chmod to change permissions of the path in case the path has already been created and provided by kubernetes